File: python/min_video.py
import math

VIDEO_SIZE_BYTE_LENGTH = 8
VIDEO_MAX_DIMENSION = VIDEO_SIZE_BYTE_LENGTH * 255
BYTES_BEFORE_FRAMES = VIDEO_SIZE_BYTE_LENGTH * 2

def dimension_split(dimension: int) -> list[int]:
    res = []

    if dimension != 0:
        count = math.ceil(dimension / 255)
        res = [dimension // count] * count

        for i in range(dimension % count):
            res[i] += 1

    while len(res) < VIDEO_SIZE_BYTE_LENGTH:
        res.append(0)

    return res

def get_coords_at_idx(frame_index: int, width: int, height: int) -> tuple[int, int]:
    x = frame_index % width
    y = (frame_index // width) % height

    return x, y


class Frame:
    def __init__(self, width, height) -> None:
        assert width <= VIDEO_MAX_DIMENSION, "Width cannot be greater than " + str(VIDEO_MAX_DIMENSION)
        assert height <= VIDEO_MAX_DIMENSION, "Height cannot be greater than " + str(VIDEO_MAX_DIMENSION)
        
        self.pixels: list[ list[int] ] = []
        self.width = width
        self.height = height

        self.pixels = [None for _ in range(width * height)]

    def get_index(self, x: int, y: int) -> int:
        return y * self.width + x

    def set_color(self, x: int, y: int, r: int, g: int, b: int):
        idx = self.get_index(x, y)
        
        assert x <= self.width, "X out of range"
        assert y <= self.height, "Y out of range"

        px = [r, g, b]
        self.pixels[idx] = px

    def get_color(self, x: int, y: int) -> list[int]:
        idx = self.get_index(x, y)

        if idx >= len(self.pixels):
            return

        col = self.pixels[idx]
        
        if not col:
            return [0, 0, 0]

        return col

    def get_data(self) -> list[int]:
        res = []

        for i in self.pixels:
            for c in i:
                res.append(c)

        return res

class Video:
    def __init__(self, width: int, height: int) -> None:
        assert width <= VIDEO_MAX_DIMENSION, "Width cannot be greater than " + str(VIDEO_MAX_DIMENSION)
        assert height <= VIDEO_MAX_DIMENSION, "Height cannot be greater than " + str(VIDEO_MAX_DIMENSION)

        self.frames: list[Frame] = []
        self.width = width
        self.height = height

    def add_frame(self, frame: Frame):
        self.frames.append(frame)

    def get_data(self) -> list[int]:
        data = []

        data.extend(dimension_split(self.width))
        data.extend(dimension_split(self.height))

        for frame in self.frames:
            for color in frame.get_data():
                data.append(color)

        return data

    def get_bytes(self) -> bytes:
        return bytes(self.get_data())

    def save_file(self, path: str):
        with open(path, "wb") as file:
            file.write(self.get_bytes())

    @staticmethod
    def get_width_from_data(data: list[int]) -> int:
        return sum(data[:VIDEO_SIZE_BYTE_LENGTH])

    @staticmethod
    def get_height_from_data(data: list[int]) -> int:
        return sum(data[VIDEO_SIZE_BYTE_LENGTH:VIDEO_SIZE_BYTE_LENGTH * 2])

    @staticmethod
    def get_frame_amount_from_data(data: list[int]) -> int:
        w = Video.get_width_from_data(data)
        h = Video.get_height_from_data(data)

        return (len(data) - VIDEO_SIZE_BYTE_LENGTH * 2) // 3 // (w * h)

    @staticmethod
    def from_data(data: list[int]):
        data_len = len(data)
        w = Video.get_width_from_data(data)
        h = Video.get_height_from_data(data)
        pixel_amt = w * h * 3
        vid = Video(w, h)
        frames = (data_len - BYTES_BEFORE_FRAMES) // pixel_amt

        for frame_i in range(frames):
            frame = Frame(w, h)

            color_start_index = BYTES_BEFORE_FRAMES + frame_i * pixel_amt
            colors = data[color_start_index:color_start_index + pixel_amt]

            for i in range(pixel_amt // 3):
                x, y = get_coords_at_idx(i, w, h)

                b = colors[i * 3]
                g = colors[i * 3 + 1]
                r = colors[i * 3 + 2]

                frame.set_color(x, y, r, g, b)

            vid.add_frame(frame)

        return vid
    
    @staticmethod
    def foreach_frame(data: list[int], callback):
        data_len = len(data)
        w = Video.get_width_from_data(data)
        h = Video.get_height_from_data(data)
        pixel_amt = w * h * 3

        frames = (data_len - BYTES_BEFORE_FRAMES) // pixel_amt

        for frame_i in range(frames):
            frame = Frame(w, h)

            color_start_index = BYTES_BEFORE_FRAMES + frame_i * pixel_amt
            colors = data[color_start_index:color_start_index + pixel_amt]

            for i in range(pixel_amt // 3):
                x, y = get_coords_at_idx(i, w, h)

                b = colors[i * 3]
                g = colors[i * 3 + 1]
                r = colors[i * 3 + 2]

                frame.set_color(x, y, r, g, b)

            callback(frame)

    @staticmethod
    def from_file(path: str):
        with open(path, "rb") as file:
            return Video.from_data(file.read())


if __name__ == "__main__":
    print("This script is to be used a library! Refer to 'cli.py'")
    print("Use `import min_video` to access its classes and functions")

File: rust/src/min_video.rs
pub const VIDEO_SIZE_BYTE_LENGTH: u32 = 8;
pub const VIDEO_MAX_DIMENSION: u32 = VIDEO_SIZE_BYTE_LENGTH * 255;
pub const BYTES_BEFORE_FRAMES: u32 = VIDEO_SIZE_BYTE_LENGTH * 2;

// Frame

pub struct Frame {
    data: Vec<u8>,
    width: u32,
    height: u32
} 

impl Frame {

    // Constructor

    pub fn new(width: u32, height: u32) -> Self {
        assert!(width > 0, "Width must be greater than 0");
        assert!(height > 0, "Height must be greater than 0");
        assert!(width <= VIDEO_MAX_DIMENSION, "Width exceeds the maximum width");
        assert!(height <= VIDEO_MAX_DIMENSION, "Width exceeds the maximum height");

        return Frame::from_data(width, height, vec![0].repeat( (width * height * 3) as usize));
    }

    pub fn from_data(width: u32, height: u32, data: Vec<u8>) -> Self {
        assert!(data.len() >= BYTES_BEFORE_FRAMES as usize, "Data too short - does not include width and height");
        assert!(data.len() == (width * height * 3) as usize, "Data length does not match width * height * 3");

        Frame {
            data: data,
            width: width,
            height: height
        }
    }

    // Dynamic functions

    pub fn set_color(&mut self, x: u32, y: u32, rgb: (u8, u8, u8)) {
        assert!(x <= self.width, "X out of range");
        assert!(y <= self.height, "Y out of range");

        let begin = get_idx_at_coords(x, y, self.width) as usize * 3;
        let (r, g, b) = rgb;

        assert!(begin + 2 < self.data.len(), "End color index out of range, this isn't supposed to happen!");

        self.data[begin + 0] = r;
        self.data[begin + 1] = g;
        self.data[begin + 2] = b;
        
    }

    pub fn get_color(&self, x: u32, y: u32) -> (u8, u8, u8) {
        assert!(x <= self.width, "X out of range");
        assert!(y <= self.height, "Y out of range");

        let begin = get_idx_at_coords(x, y, self.width) as usize * 3;

        // TEMPORARY FIX
        // TODO: Find the cause why the colors are inverted
        let b = self.data[begin + 0];
        let g = self.data[begin + 1];
        let r = self.data[begin + 2];

        return (r, g, b);
    }

    pub fn get_data(&self) -> &Vec<u8> {
        return &self.data;
    }

}

// Video class

pub struct Video {
    data: Vec<u8>,
    width: u32,
    height: u32
}

impl Video {

    // Constructor

    pub fn new(width: u32, height: u32) -> Self {
        assert!(width > 0, "Width must be greater than 0");
        assert!(height > 0, "Height must be greater than 0");
        
        Video {
            data: dimension_split(width).iter().chain(dimension_split(height).iter()).cloned().collect(),
            width: width,
            height: height
        }
    }

    /// Constructs a video from a data buffer
    pub fn from_data(data: &Vec<u8>) -> Self {
        assert!(Video::is_data_valid(data), "Not a valid MinVideo data");

        let w = Video::get_width_from_data(data);
        let h = Video::get_height_from_data(data);

        assert!(w > 0, "Data invalid, width is 0");
        assert!(h > 0, "Data invalid, width is 0");
        
        Video {
            data: data.to_vec(),
            width: w,
            height: h
        }
    }

    // Dynamic functions

    /// Adds a frame to the video
    pub fn add_frame(&mut self, frame: &Frame) {
        assert!(frame.width == self.width, "Frame width not equal to video width");
        assert!(frame.height == self.height, "Frame height not equal to video height");

        self.data.extend(frame.data.iter());
    }

    /// Returns the frame at the specified index
    pub fn get_frame(&mut self, index: usize) -> Frame {
        let begin = BYTES_BEFORE_FRAMES + (self.width * self.height * 3) * index as u32;
        let end = begin + (self.width * self.height * 3);

        let frame_data = self.data[begin as usize..end as usize].to_vec();
        return Frame::from_data(self.width, self.height, frame_data);
    }

    /// Returns the video's data buffer
    pub fn get_data(&self) -> Vec<u8> {
        return self.data.clone();
    }

    /// Returns the amount of frames in the video
    pub fn get_frame_amount(&self) -> usize {
        return Video::get_frame_amount_from_data(&self.data);
    }

    /// Returns the video width
    pub fn get_width(&self) -> u32 {
        return self.width;
    }

    /// Returns the video height
    pub fn get_height(&self) -> u32 {
        return self.height;
    }

    // Static functions

    /// Checks if the specified data can be used to construct a video
    pub fn is_data_valid(data: &Vec<u8>) -> bool {
        return data.len() >= BYTES_BEFORE_FRAMES as usize;
    }

    /// Calculates the width from a video data buffer
    pub fn get_width_from_data(data: &Vec<u8>) -> u32 {
        let mut res: u32 = 0;

        for i in data.iter().take(VIDEO_SIZE_BYTE_LENGTH as usize) {
            res += *i as u32;
        }

        return res;
    }

    /// Calculates the height from a video data buffer
    pub fn get_height_from_data(data: &Vec<u8>) -> u32 {
        let mut res: u32 = 0;

        for i in data.iter().skip(VIDEO_SIZE_BYTE_LENGTH as usize).take(VIDEO_SIZE_BYTE_LENGTH as usize) {
            res += *i as u32;
        }

        return res;
    }

    /// Calculates the amount of frames from a video buffer
    pub fn get_frame_amount_from_data(data: &Vec<u8>) -> usize {
        let w = Video::get_width_from_data(data);
        let h = Video::get_height_from_data(data);
        
        return (data.len() - BYTES_BEFORE_FRAMES as usize) / 3 / (w * h) as usize;
    }
}

// Functions

pub fn dimension_split(dimension: u32) -> Vec<u8> {
    let mut res: Vec<u8> = Vec::new();

    if dimension != 0 {
        let count = (dimension as f64 / 255.0).ceil() as u32;

        for _ in 0..count {
            res.push( (dimension / count) as u8 );
        }

        for i in 0.. dimension % count {
            res[i as usize] += 1;
        }
    }

    while res.len() < VIDEO_SIZE_BYTE_LENGTH as usize {
        res.push(0);
    }

    return res;
} 

pub fn get_coords_at_idx(index: u32, width: u32, height: u32) -> (u32, u32) {
    let x = index % width;
    let y = (index / width) % height;

    return (x, y)
}

pub fn get_idx_at_coords(x: u32, y: u32, width: u32) -> u32 {
    return y * width + x;
}